File: frontend/src/services/rent.service.ts
import moment from "moment";
import { DateFormats } from "../models/enums/date";
import { HomeDetailsModalPropsI } from "../models/interfaces/home-details-modal.props";
import { RentContactTypes } from "../models/interfaces/rent-contact-types";
import { FormatterService } from "./formatter.service";
import { TimeService } from "./time.service";
import { toast } from "react-toastify";

export class RentService {
  calcFinalValue(props: {
    startDate: Date;
    endDate: Date;
    pricePerNight: number | undefined;
  }) {
    if (!props.startDate || !props.endDate) {
      return "Selecione as datas para calcular";
    }

    const _timeService = new TimeService();
    const _formatterService = new FormatterService();

    let daysCount = _timeService.getDaysDiff(props.startDate, props.endDate);
    let totalPrice = props.pricePerNight! * daysCount;
    return _formatterService.toBRL(totalPrice);
  }

  sendMessage(type: RentContactTypes, props: HomeDetailsModalPropsI) {
    if (
      !props.name ||
      !props.startDate ||
      !props.endDate ||
      !props.guestCount
    ) {
      return;
    }

    let amount = this.calcFinalValue({
      startDate: new Date(props.startDate),
      endDate: new Date(props.endDate),
      pricePerNight: props.house!.pricePerNight,
    });
    props.setFinalValue(amount);

    let message = `
    Olá ${props.house?.responsible.name}, me chamo ${props.name}.\n
    Tenho interesse em alugar o imóvel *${props.house!.name} nos dias ${moment(
      props.startDate
    ).format(DateFormats.DATE)} até ${moment(props.endDate).format(
      DateFormats.DATE
    )}.\n
    Estarei com ${props.guestCount} pessoas.\n
    O valor final calculado é ${amount}.\n
    Além disso gostaria de deixar essas observações: ${props.notes}`;

    console.log("Here");
    if (type === RentContactTypes.EMAIL) {
      window.location.href = `mailto:${props.house?.responsible.email}?subject=Aluguel de imóvel&body=${message}`;
      return;
    }
    if (type === RentContactTypes.PHONE) {
      window.location.href = `https://wa.me//${props.house?.responsible.phone}?text=${message}`;
      return;
    }
  }
}

File: mobile/lib/services/rent.service.dart
import 'package:stay_floripa/common/currency.dart';
import 'package:stay_floripa/common/date.dart';
import 'package:stay_floripa/models/house_model.dart';
import 'package:flutter/material.dart';
import 'package:url_launcher/url_launcher_string.dart';

class SendRentMessageModel {
  final HouseDetailsModel house;
  final DateTimeRange dateRange;
  final double amount;
  final int guestCount;
  final String notes;
  final String name;

  SendRentMessageModel({
    required this.house,
    required this.dateRange,
    required this.amount,
    required this.guestCount,
    required this.notes,
    required this.name,
  });
}

class RentService {
  String message = "";

  String _buildMessage(SendRentMessageModel props) {
    message =
        "Olá ${props.house.contact?.name}, me chamo ${props.name}. Gostaria de alugar o imóvel ${props.house.title} nos dias ${formatDate(props.dateRange.start)} até ${formatDate(props.dateRange.end)}. Estarei com ${props.guestCount} pessoas. O aplicativo calculou um valor total de ${getAmountFormatted(props.amount)}. ";

    if (props.notes.isNotEmpty) {
      message +=
          'Além disso gostaria de deixar essas observações: ${props.notes}';
    }

    return message;
  }

  Future<void> _launchURL(String url) async {
    if (await canLaunchUrlString(url)) {
      await launchUrlString(url);
    } else {
      throw 'Não foi possível abrir o link $url';
    }
  }

  void sendWhatsApp(SendRentMessageModel props) async {
    _buildMessage(props);
    await _launchURL(
      'https://wa.me//${props.house.contact!.phone}?text=$message',
    );
  }

  void sendEmail(
    SendRentMessageModel props,
  ) async {
    _buildMessage(props);
    await _launchURL(
      "mailto:${props.house.contact!.email}?subject=Aluguel de imóvel&body=$message",
    );
  }
}
